DOCS-#2996: document `modin.backends.base` (#3021)
modin-project/modin: docs/flow/modin/backends/index.rst

:orphan:

Query compiler
==============

> Modin has several execution backends. Calling any DataFrame API function will end up in

Reviewer: ```suggestion
Modin supports several execution backends. Calling any DataFrame API function will end up in
```

> some backend-specific method. Query compiler is a bridge between Modin DataFrame and

Reviewer: ```suggestion
some backend-specific method. The query compiler is a bridge between Modin Dataframe and
```

an actual execution engine.

.. image:: /img/simplified_query_flow.svg
    :align: right
    :width: 300px

> Query compilers of all backends provide a common API, which is used by Modin DataFrame

Reviewer: ```suggestion
Query compilers of all backends implement a common API, which is used by the Modin Dataframe
```

> to emulate pandas API. The role of the query compiler is to translate its API into

Reviewer: ```suggestion
to support dataframe queries. The role of the query compiler is to translate its API into
```

Author: Does DataFrame uses a query compiler to support "dataframe queries"? What are "dataframe queries" and why we should use this term instead of "pandas API"?

Reviewer: The Modin implementation is more general than just pandas. Since this is developer documentation, we can assume a higher technical level and talk more generally.

> a set of DataFrame algebra operations. Each query compiler instance contains

Reviewer: ```suggestion
a pairing of known user-defined functions and dataframe algebra operators. Each query compiler instance contains a
```

> :doc:`frame </flow/modin/engines/base/frame/index>` of the selected execution engine and query

Reviewer: ```suggestion
:doc:`frame </flow/modin/engines/base/frame/index>` of the selected execution engine and queries
```

it with the compiled queries to get the result. The query compiler object is immutable,
so the result of every method is a new query compiler.

> Query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class

Reviewer: ```suggestion
The query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
```

> and partially mimic the pandas API, however, they're not equal. Query compilers API

Reviewer: ```suggestion
and may resemble the pandas API, however, they're not equal. The query compilers API
```

is significantly reduced in comparison with pandas, since many corner cases or even the

> whole methods can be handled at the DataFrame level with the existing API.

Reviewer: ```suggestion
whole methods can be handled at the API layer with the existing API.
```

> Query compiler is the level where Modin stops distinguishing Frame and Series objects.

Reviewer: ```suggestion
The query compiler is the level where Modin stops distinguishing Frame and Series (or column) objects.
```

Author: just a bit of curiosity, what does "Series **or column**" means?

Reviewer 2: I believe @devin-petersohn mean that we treat Series (or column - column of a DataFrame is a Series object as well) as DataFrame at the QC layer.

> Series is represented by one-column query compiler, where Series name is the column label,

Reviewer: ```suggestion
A Series is represented by a one-column query compiler, where the Series name is the column label.
```

> if Series is unnamed, then the label would be ``"__reduced__"``. DataFrame API level

Reviewer: ```suggestion
If Series is unnamed, then the label is ``"__reduced__"``. The Dataframe API layer
```

> interprets one-column query compilers as Series or DataFrame depending on the operation context.

Reviewer: ```suggestion
interprets a one-column query compiler as Series or DataFrame depending on the operation context.
```

Query Compiler
==============

.. toctree::
    :hidden:

    base/query_compiler
    pandas/index
    cudf/index
    pyarrow/index

Modin supports several execution backends. Calling any DataFrame API function will end up in
some backend-specific method. The query compiler is a bridge between Modin Dataframe and
an actual execution engine.

.. image:: /img/simplified_query_flow.svg
    :align: right
    :width: 300px

Query compilers of all backends implement a common API, which is used by the Modin Dataframe
to emulate pandas API. The role of the query compiler is to translate its API into
a pairing of known user-defined functions and dataframe algebra operators. Each query compiler instance contains a
:doc:`frame </flow/modin/engines/base/frame/index>` of the selected execution engine and queries
it with the compiled queries to get the result. The query compiler object is immutable,
so the result of every method is a new query compiler.

> The query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class

Reviewer 3: ```suggestion
The query compilers API is defined by the :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` class
```
It generates auto-link to `BaseQueryCompiler` class API in other .rst file

The query compiler is the level where Modin stops distinguishing Frame and Series (or column) objects.
A Series is represented by a one-column query compiler, where the Series name is the column label.
If Series is unnamed, then the label is ``"__reduced__"``. The Dataframe API layer
interprets a one-column query compiler as Series or DataFrame depending on the operation context.

.. note::
    Although we're declaring that there is no difference between DataFrame and Series at the query compiler,
    you still may find methods like ``method_ser`` and ``method_df`` which perform differently because they
    emulating either Series or DataFrame logic, or you may find parameters, which indicates whether this one-column
    query compiler is representing Series or not. All of these are hacks, and we're working on getting rid of them.

High-level module overview
''''''''''''''''''''''''''
This module houses submodules of all of the stable query compilers:

- :doc:`Base module<base/query_compiler>` contains an abstract query compiler class which defines common API.
- :doc:`Pandas module<pandas/index>` contains query compiler and text parsers for pandas backend.

> - :doc:`Cudf module<cudf/index>` contains query compiler and text parsers for Cudf backend.

Reviewer 3: Suggestion above wasn't applied, but resolved. Can you explain?

Author: Can you please refer to that suggestion? I thought that I applied all of them

Reviewer 3: It's here https://github.com/modin-project/modin/pull/3021#discussion_r631825296

Author: Oh, thanks. Fixed that in the last commit as well.
